graphql_compiler/macros/macro_edge/validation.py

```diff
@@ -0,0 +1,113 @@
+                                of every GraphQL type in the "value" GraphQL union.
+                                Recursive expansion of type equivalence hints is not performed,
+                                and only type-level correctness of this argument is enforced.
+                                See README.md for more details on everything this parameter does.
+                                *****
+                                Be very careful with this option, as bad input here will
+                                lead to incorrect output queries being generated.
+                                *****
+
+    Returns:
+        tuple (class name for macro, name of macro edge, MacroEdgeDescriptor),
+        where the first two values are strings and the last one is a MacroEdgeDescriptor object
+    """
+    if ast.directives is not None:
+        directive_names = [directive.name.value for directive in ast.directives]
+        raise GraphQLInvalidMacroError(
+            u'Unexpectedly found directives at the top level of the GraphQL input. '
+            u'This is not supported. Directives: {}'.format(directive_names))
+
+    if ast.variable_definitions is not None:
+        raise GraphQLInvalidMacroError(
+            u'Unexpectedly found variable definitions at the top level of the GraphQL input. '
+            u'This is not supported. Variable definitions: {}'.format(ast.variable_definitions))
+
+    unique_and_parameterless_directives_to_check = (
+        MacroEdgeDirective,
+        MacroEdgeDefinitionDirective,
+        MacroEdgeTargetDirective,
+    )
+    macro_edge_sub_asts = {}
+
+    for directive_definition in unique_and_parameterless_directives_to_check:
+        macro_data = macro_directives.get(directive_definition.name, None)
+        if not macro_data:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" was not found anywhere within the supplied '
+                u'macro edge definition GraphQL.'.format(directive_definition.name))
+
+        if len(macro_data) > 1:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" was unexpectedly present more than once in '
+                u'the supplied macro edge definition GraphQL. It was found {} times.'
+                .format(directive_definition.name, len(macro_data)))
+
+        macro_ast, macro_directive = macro_data[0]
+        if macro_directive.arguments is not None:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" unexpectedly contained arguments even though '
+                u'it is not supposed to contain any. Unexpected arguments:'
+                .format(directive_definition.name, macro_directive.arguments))
+
+        macro_edge_sub_asts[directive_definition.name] = macro_ast
+
+    # TODO(predrag): Required further validation:
+    # - the target directive AST is within the definition directive AST;
```

Comment:
Aren't they also allowed to be on the same place?

Reply:
Fixed, good catch.